dashboard/src/actions/datasetListActions.js

```diff
@@ -1,27 +1,74 @@
 import * as TYPES from "./types";
 
 import API from "../utils/axiosInstance";
+import { DANGER } from "assets/constants/toastConstants";
+import { showToast } from "./toastActions";
 import { uriTemplate } from "utils/helper";
 
-export const fetchPublicDatasets = () => async (dispatch, getState) => {
+export const fetchPublicDatasets = (page) => async (dispatch, getState) => {
   try {
     dispatch({ type: TYPES.LOADING });
     const endpoints = getState().apiEndpoint.endpoints;
+    const { offset, limit, filter, searchKey, perPage } =
+      getState().datasetlist;
+    let publicData = [...getState().datasetlist.publicData];
+    const params = new URLSearchParams();
+    params.append("metadata", "dataset.uploaded");
+    params.append("access", "public");
+    params.append("offset", offset);
+    params.append("limit", limit);
+
+    if (searchKey) {
+      params.append("name", searchKey);
+    }
+    if (filter.startDate instanceof Date && !isNaN(filter.startDate)) {
+      params.append("start", new Date(filter.startDate).toUTCString());
+    }
+    if (filter.endDate instanceof Date && !isNaN(filter.endDate)) {
+      params.append("end", new Date(filter.endDate).toUTCString());
+    }
+
     const response = await API.get(
       uriTemplate(endpoints, "datasets_list", {}),
-      { params: { metadata: "dataset.uploaded", access: "public" } }
+      { params }
     );
+
     if (response.status === 200 && response.data) {
+      const startIdx = (page - 1) * perPage;
+
+      if (publicData.length === 0) {
+        publicData = new Array(response.data.total);
+      }
+      publicData.splice(
+        startIdx,
+        response.data.results.length,
+        ...response.data.results
+      );
+
+      dispatch({
+        type: TYPES.GET_PUBLIC_DATASETS,
+        payload: [...publicData],
+      });
+      // in case of last page, next_url is empty
+      if (response.data.next_url) {
+        const urlSearchParams = new URLSearchParams(response.data.next_url);
+        const params = Object.fromEntries(urlSearchParams.entries());
+
+        dispatch({
+          type: TYPES.SET_PAGE_OFFSET,
+          payload: Number(params?.offset),
+        });
+      }
       dispatch({
-        type: "GET_PUBLIC_DATASETS",
-        payload: response?.data?.results,
+        type: TYPES.SET_TOTAL_DATASETS,
+        payload: response.data.total,
       });
```

Comment:
As _totalDatasets_ is got from the response, it's being maintained separately.

Reply:
So, we get the response, we extract the `total` value from it, we use that value to set the size of the `publicData` array, and then we store both the array and the `total` value in the Redux store (right?).

Do we ever use the stored value of `total`?  (I'm not seeing it right off the bat.)  And, do we ever use it in a context where we couldn't easily just use `publicData.length` instead?

(I'm not seeing any value in storing `total` at the moment.)

Reply:
`total` is being used in the Pagination component

Reply:
Ah!  Thank you.  (I was looking for the wrong name.)
```javascript
  const { totalDatasets, publicData, perPage } = useSelector(
```
So, yes, where we are pulling in `totalDatasets` we are also pulling in `publicData`, so we could just use `publicData.length` instead of `totalDatasets`, right?

Reply:
In fact, at `PaginationComponent/index.jsx` line 61, just a dozen lines before the reference to `totalDatasets`, we have a reference to `publicData.length`.  🙂

Reply:
Replaces references to  `totalDatasets`  with `publicData.length`